### PR TITLE
Add longhorn storage test

### PR DESCRIPTION
**Pull request by brooksn**
#### Proposed Changes ####

Adds a basic test for the Longhorn setup steps in the k3s documentation:
https://docs.k3s.io/storage#setting-up-longhorn

#### Linked Issues ####

#6061 

#### User-Facing Change ####

```release-note

```

### CODE REVIEW

File: tests/integration/longhorn/longhorn_int_test.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+var server *testutil.K3sServer
+var serverArgs = []string{"--cluster-init"}
+var testLock int
+
+var _ = BeforeSuite(func() {
+	if _, err := exec.LookPath("iscsiadm"); err != nil {
+		Skip("Test needs open-iscsi to ben installed")
```
Comment (brandond) [inline tests/integration/longhorn/longhorn_int_test.go:24]:
```suggestion
		Skip("Test needs open-iscsi to be installed")
```

**Review by brandond — COMMENTED** [review]
one nit, LGTM otherwise!

Do we want to also test the CNI snapshotter functionality? That requires installing some extra bits I guess...

File: tests/integration/longhorn/longhorn_int_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	config, err := clientcmd.BuildConfigFromFlags("", "/etc/rancher/k3s/k3s.yaml")
+	if err != nil {
+		return nil, err
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
```
Comment (dereknola) [inline tests/integration/longhorn/longhorn_int_test.go:153]: This function is a duplicate of the existing `testutils.k8sClient` function. Just export that for use here.

**Review by dereknola — CHANGES_REQUESTED** [review]

**Comment by dereknola** [comment]
Looking at the CI failure, you may need to A) bump the timeout time on some of your checks or B) reduce the request size of the storage. The GH runners don't have that much space on them.

**Comment by matttrach** [comment]
just to short circuit some troubleshooting attempts the following error was thrown:
```
• [FAILED] [30.386 seconds]
longhorn
/home/runner/work/k3s/k3s/tests/integration/longhorn/longhorn_int_test.go:35
  when persistent volume claim is created
  /home/runner/work/k3s/k3s/tests/integration/longhorn/longhorn_int_test.go:84
    [It] creates a pod with the pvc
    /home/runner/work/k3s/k3s/tests/integration/longhorn/longhorn_int_test.go:111

  Timed out after 30.004s.
  Expected success, but got an error:
      <*errors.errorString | 0xc0003b2370>: {
          s: "pod volume-test not running",
      }
      pod volume-test not running
  In [It] at: /home/runner/work/k3s/k3s/tests/integration/longhorn/longhorn_int_test.go:128
  ```

**Comment by matttrach** [comment]
The pvc was successfully generated in the previous test: 
```
• [SLOW TEST] [11.487 seconds]
longhorn when persistent volume claim is created creates the pv and pvc
/home/runner/work/k3s/k3s/tests/integration/longhorn/longhorn_int_test.go:85
```

**Comment by matttrach** [comment]
Looking at the debug log it appears to have timed out attempting to ssh into the container:
```
SSH: ssh 3spcPn8DJjWWr9d6YACQAky9J@sfo2.tmate.io
Web shell: https://tmate.io/t/3spcPn8DJjWWr9d6YACQAky9J
```

It says "web shell", does that mean it is attempting on port 80? 
That is the only port exposed by the pod, I believe.

File: tests/integration/longhorn/longhorn_int_test.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+				if err != nil {
+					return fmt.Errorf("failed to get pod volume-test")
+				}
+				if pod.Status.Phase != "Running" {
+					return fmt.Errorf("pod volume-test \"%s\" reason: \"%s\" message \"%s\"", pod.Status.Phase, pod.Status.Reason, pod.Status.Message)
+				}
+				return nil
+			}, "60s", "5s").Should(Succeed())
+		})
+	})
```
Comment (dereknola) [inline tests/integration/longhorn/longhorn_int_test.go:112]: Can you add a section deletion the longhorn pvc? Also doubles as cleanup for test.

**Review by dereknola — CHANGES_REQUESTED** [review]

File: tests/integration/integration.go [hunk 4]
```diff
@@ -188,6 +188,42 @@ func ParseNodes() ([]corev1.Node, error) {
 	return nodes.Items, nil
 }
 
+func GetPod(namespace, name string) (*corev1.Pod, error) {
+	client, err := k8sClient()
+	if err != nil {
+		return nil, err
+	}
+	pod, err := client.CoreV1().Pods(namespace).Get(context.Background(), name, metav1.GetOptions{})
```
Comment (briandowns) [inline tests/integration/integration.go:196]: You can return here.

File: tests/integration/integration.go [hunk 5]
```diff
@@ -188,6 +188,42 @@ func ParseNodes() ([]corev1.Node, error) {
...
+	}
+	return pod, nil
+}
+
+func GetPersistentVolumeClaim(namespace, name string) (*corev1.PersistentVolumeClaim, error) {
+	client, err := k8sClient()
+	if err != nil {
+		return nil, err
+	}
+	pvc, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(context.Background(), name, metav1.GetOptions{})
```
Comment (briandowns) [inline tests/integration/integration.go:208]: You can return here.

File: tests/integration/integration.go [hunk 6]
```diff
@@ -188,6 +188,42 @@ func ParseNodes() ([]corev1.Node, error) {
...
+	}
+	return pvc, nil
+}
+
+func GetPersistentVolume(name string) (*corev1.PersistentVolume, error) {
+	client, err := k8sClient()
+	if err != nil {
+		return nil, err
+	}
+	pv, err := client.CoreV1().PersistentVolumes().Get(context.Background(), name, metav1.GetOptions{})
```
Comment (briandowns) [inline tests/integration/integration.go:220]: You can return here.

File: tests/integration/longhorn/longhorn_int_test.go (hunk 3 above)
Comment (brooksn) [reply tests/integration/longhorn/longhorn_int_test.go:112]: Yep, it looks like this test cleans up the storage replicas.

**Review by dereknola — APPROVED** [review]

File: tests/integration/longhorn/longhorn_int_test.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var server *testutil.K3sServer
+var serverArgs = []string{"--cluster-init"}
+var testLock int
+
+var _ = BeforeSuite(func() {
+	if _, err := exec.LookPath("iscsiadm"); err != nil {
+		Skip("Test needs open-iscsi to be installed")
```
Comment (exp0nge) [inline tests/integration/longhorn/longhorn_int_test.go:21]: What setups this up?

Comment (brandond) [reply tests/integration/longhorn/longhorn_int_test.go:21]: Please open a discussion if you have questions. Year old PRs are not the correct place.

**Pull request merged**